internal/k8s/service/resolver.go

```diff
@@ -342,9 +341,83 @@ func (r *backendResolver) findGlobalCatalogService(service *corev1.Service) (*Re
 	return nil, nil
 }
 
-// this will resolve a service based on our future CRD
-func (r *backendResolver) findConsulMeshService(name types.NamespacedName) {
-	mapped := r.mapper(name.Namespace)
+func (r *backendResolver) consulServiceForMeshService(ctx context.Context, namespacedName types.NamespacedName) (*ResolvedReference, error) {
+	var err error
+	var resolved *ResolvedReference
+
+	service, err := r.client.GetMeshService(ctx, namespacedName)
+	if err != nil {
+		r.logger.Trace("error retrieving mesh service", "error", err, "name", namespacedName.Name, "namespace", namespacedName.Namespace)
+		return nil, err
+	}
+	if service == nil {
+		return nil,
+			NewConsulResolutionError(fmt.Sprintf(
+				"kubernetes mesh service object (%q, %q) not found", namespacedName.Name, namespacedName.Namespace,
+			))
+	}
+
+	// we do an inner retry since consul may take some time to sync
+	err = backoff.Retry(func() error {
+		r.logger.Trace("attempting to resolve global catalog service")
+		resolved, err = r.findCatalogService(service)
+		if err != nil {
+			r.logger.Trace("error resolving global catalog reference", "error", err)
+			return err
+		}
+		if resolved == nil {
+			return NewConsulResolutionError(fmt.Sprintf("consul service %s not found", namespacedName))
+		}
+		return nil
+	}, backoff.WithContext(backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), 30), ctx))
+	if err != nil {
+		r.logger.Error("could not resolve consul service", "error", err)
+		return nil, err
+	}
+	return resolved, nil
+}
 
-	r.logger.Debug("resolving consul mesh service", "name", name.Name, "namespace", mapped)
+func (r *backendResolver) findCatalogService(service *apigwv1alpha1.MeshService) (*ResolvedReference, error) {
+	consulNamespace := r.mapper(service.Namespace)
+	consulName := service.Spec.Name
+	services, _, err := r.consul.Catalog().Service(consulName, "", &api.QueryOptions{
+		Namespace: consulNamespace,
+	})
+	if err != nil {
+		r.logger.Trace("error resolving consul service reference", "error", err)
+		return nil, err
+	}
+	if len(services) == 0 {
+		return nil, NewConsulResolutionError(fmt.Sprintf("consul service (%s, %s) not found", consulNamespace, consulName))
```

Comment:
Since we're pulling the namespace from the mapping code and the name from the resolved `MeshService` parameter, I'd prefer to avoid an additional struct initialization just for formatting sake -- that said, might be worth doing a follow-up PR going through this file and the rest of the codebase to settle on some standardized logging and error messaging formats.